WIP - fix(dataView): add option to apply row selection to all pages

ghiscoding opened the pull request:
- when using local Pagination and DataView, it was applying row selection only to current page but most users would expect the selection to be applied to all pages, to fix that we added:
   -  new CheckboxSelectColumn plugin option `applySelectOnAllPages` was added, its default is `false` so it won't affect previous grids
   - add few more DataView functions
      - `getItemsCount` for an easy access to items length
      - `setAllSelectedIds` which is called by CheckboxSelectColumn plugin `applySelectOnAllPages` is set to true
      - `getAllSelectedFilteredIds` and `getAllSelectedFilteredItems`  to get all filtered data when using Pagination `getAllSelectedFilteredIds` and `getAllSelectedFilteredItems`
- fixes issue identified and discussed in this issue [comment](https://github.com/6pac/SlickGrid/pull/542#issuecomment-709557314)

#### TODOs
- [x] modify `example-checkbox-header-row.html` to demo the feature with `applySelectOnAllPages`
- [x] test with `selectableOverride` callback
- [x] add Cypress E2E tests and fix failing tests
  - [x] add test to make sure "Select All" gets unchecked when unselecting any row
- [ ] `selectedIds` should be updated after filtering data (we should use `filteredItems`)

#### WITH FIX
![tePoBbFn9a](https://user-images.githubusercontent.com/643976/96602324-e6322980-12c0-11eb-9ff5-169c1a274f40.gif)

#### without fix
![L0kahu5ej5](https://user-images.githubusercontent.com/643976/96475913-92610b00-1202-11eb-9216-1c904f04ad37.gif)

ghiscoding commented:
@arashdalir @6pac
I managed to get it working and I modified `example-checkbox-header-row.html` to test it out (demoed in the animate gif on top with/without the fix). 
**Note** that it broke some of my older Cypress E2E tests, I'll fix them soon and will add more tests to cover the new feature.

@arashdalir
If you could also test it out on your side, and provide feedback, that would be great since a double validation and a pair of eyes is always better than one 😉 

From what I tested with, it works really great and I will enable this **only** with Local Pagination (in-memory data). What I mean is that I won't enable this feature when using a grid with backend services since we don't know what's on next page until displaying it and so we shouldn't enable it (so the fact that I added a new `applySelectOnAllPages` even helps for certain scenarios like this one), for example in Angular-Slickgrid I added 2 backend services OData/GraphQL.

arashdalir commented:
> `getAllSelectedFilteredIds` and `getAllSelectedFilteredItems` to get all filtered data when using Pagination 

I cannot remember this, is there any function to get all the filtered rows (regardless if they are selected or not)?

ghiscoding commented:
> > `getAllSelectedFilteredIds` and `getAllSelectedFilteredItems` to get all filtered data when using Pagination
> 
> I cannot remember this, is there any function to get all the filtered rows (regardless if they are selected or not)?

Yes there is `getFilteredItems()` which reminds me that the issue I found on (Edit 4) is because I should use `getAllSelectedFilteredIds` instead of `getAllSelectedIds` inside the CheckboxSelectColumn plugin... was that why you asked the question?

arashdalir commented:
> > > `getAllSelectedFilteredIds` and `getAllSelectedFilteredItems` to get all filtered data when using Pagination
> > 
> > 
> > I cannot remember this, is there any function to get all the filtered rows (regardless if they are selected or not)?
> 
> Yes there is `getFilteredItems()` which reminds me that the issue I found on (Edit 4) is because I should use `getAllSelectedFilteredIds` instead of `getAllSelectedIds` inside the CheckboxSelectColumn plugin... was that why you asked the question?

well, not actually! :D but good that we realized there is an issue!
I was not able to find if there is a function which would deliver this dataset; wanted to suggest we should also add another function for that...

ghiscoding commented:
So `getFilteredItems` should do it. Have you had a chance to try out the code? It'd be nice to get your feedback.

arashdalir commented:
> So `getFilteredItems` should do it. Have you had a chance to try out the code? It'd be nice to get your feedback.

I am trying to get to that point! unfortunately, as I am integrating `slick-grid` into our product and we - as always and like any other company :D - are under pressure to deliver the product in time; so I have a bit of time stress. There are a few issues with the configurations on other parts of `slick-grid` - atm specifically with the column-auto-sizing - and as soon as I get them up and running, I'll be able to go back to this original issue... sorry!

ghiscoding commented:
Yeah I know the feeling, I'm under pressure too and many of these improvements are done in overtime (sadly). 
For the column sizing, you maybe want to read the [Column Auto Sizing](https://github.com/6pac/SlickGrid/wiki/Auto-Column-Sizing) Wiki that 6pac created.

arashdalir commented:
> Yeah I know the feeling, I'm under pressure too and many of these improvements are done in overtime (sadly).
> For the column sizing, you maybe want to read the [Column Auto Sizing](https://github.com/6pac/SlickGrid/wiki/Auto-Column-Sizing) Wiki that 6pac created.

thanks for the hint! I read it but have still not been able to find the issue in my config! I have some columns which are hidden by default, and they only have `width`, and I expect them to be shown correctly - e.g. their width is calculated correctly - after one selects them and I call the `grid.autosizeColumns()`! but this doesn't work and column widths are set to a minimum value - their widths and that of already visible columns are not adjusted. Anyways, I am working on that issue :) thanks for your help!

6pac commented:
Hi @arashdalir, yes the Autosize behaviour is a bit complicated. You can use Legacy mode to go back to the 'old' Slickgrid way of doing it, if that works better for you. When I rewrote the behaviour, I was trying for something better, but it takes a bit of tweaking and metadata to make it work well. I found particularly when there are complex column contents like dropdowns or hyperlinks it needs customised behaviour for those columns.

arashdalir commented:
hi @6pac 
thanks for your input and feedback :) I did exactly as you said and it's working for the moment, though we could check that issue separately maybe! :)
I will try to check the functionality of this PR and will get back to both of you asap...

arashdalir commented:
@ghiscoding 
I just checked the functionality; I believe there is something wrong here. Maybe my configs, but need some clarification please! apparently, the `select all` checkbox ignores my filtering!
* I have a 20-page list, I can now set `applySelectOnAllPages` and the checkbox now selects all rows on all pages! great! 🆗 
* now, I filter the values; I have 11 entries, all on 1 page. I click the `select all` button; expecting only the 11 entries to be selected. then I remove the filter; expecting to get only those 11 entries marked. but now all the 20 pages (500 entries) are selected. 🚫 

have I missed a config? below is a sample of the issue as a GIF. sorry, too long, but hopefully shows the issue...

![select-all-issue](https://user-images.githubusercontent.com/2896170/96870318-043e8c00-1471-11eb-951c-12db5256d748.gif)

ghiscoding commented:
Yes I know that bug, I wrote it in (Edit 4)

> EDIT 4 Found yet another 2 issues
when filtering data it doesn't change the Selected Ids and that is wrong (when using regular grid without Pagination it does exclude filtered item from the selection returned by grid.getSelectedRows(). So we should use the filteredItems to refresh the selectedIds accordingly.
all this work (feature) should only be used when the grid uses Pagination, if there's no Pagination then there's no point in using this new code... so I need to detect if the grid uses Pagination before enabling the feat.

and then I said 
> So `getFilteredItems` should do it. Have you had a chance to try out the code? It'd be nice to get your feedback.

which is the message that you replied too... but I didn't have a chance to do that change yet, maybe today if I get some free time

arashdalir commented:
> Yes I know that bug, I wrote it in (Edit 4)

sorry, missed that. the ticket is getting long and I am losing track of what's what tbh!

> You can maybe try yourself, try changing slick.checkboxselectcolumn.js on line 162 from this

didn't help. I will check it out and see if I can get it right...

ghiscoding commented:
Yes I tried too and that didn't work, I think it's because `selectedRowIds` doesn't get updated after items are filtered. I think at the end of it I'll remove `getAllSelectedFilteredIds` and just keep `getAllSelectedIds` since I don't see any value in keeping a reference to the unfiltered ones. If we can fix the `selectedRowIds` to be the filtered Ids then `getAllSelectedIds` will become the filtered ones anyway.

arashdalir commented:
I believe I know what's happening here; am doing some debugging and will report back. but just fyi: 
the way it's programmed now, it practically stores all 500 entries as `selectedRowIds` in `DataView`, and then delivers only those "visible" when returning the `getAllSelectedFilteredIds()`. it, instead, should only store my 14 selected rows...

![image](https://user-images.githubusercontent.com/2896170/96876556-e412cb00-1478-11eb-805f-c85456e4ad99.png)

arashdalir commented:
okay, I found the root of this problem, but now we have another one!

so, the solution is to use `_dataView.getFilteredItems()` when getting the list of items in `checkboxselectcolumn` like in the following change:

![image](https://user-images.githubusercontent.com/2896170/96880431-7321e200-147d-11eb-8112-9d3921decb19.png)

now about the new issue:
as we know, HTML's checkbox only has 2 modes: `checked` and `not checked`! because of that and based on how this plugin is written, when I filter my dataset and then push the `select-all` button, the checkbox resets to `unchecked`, as the code controlling it realizes there are 500 rows and only 14 of them are selected; and sets the `select-all`  to `unchecked`. This means that I cannot use the checkbox to unselect the filtered rows anymore (because the state before my click is always `unchecked` and after my click always `checked`)!
not sure if I could explain the issue correctly. maybe this GIF would help visualising the issue. we need to add the third state as @6pac  mentioned...

![select-all-issue2](https://user-images.githubusercontent.com/2896170/96881393-84b7b980-147e-11eb-90e2-278430af8129.gif)

arashdalir commented:
I mean, we need to implement the 3rd `partially-checked` mode to allow `unselecting` the filtered rows...

arashdalir commented:
damn! wait please! I have another bug too; the concatenation is not working in dataview; I will post the final solution as soon as I get it...

arashdalir commented:
okay, this change is also necessary in `dataview`. if I understand it correctly, notifying the `onSelectedRowIdsChanged` will cause the event `onSelectedRowsChanged` to be triggered, and there the function `setSelectedRowIds` will be called, which again calls `onSelectedRowsChanged` with **all the selected rows** (those selected before and those selected now); which calls `setSelectedRowIds` again, but this time, the recursion breaker ` if (selectedRowIds.join(",") == rowIds.join(","))` is true and therefore the recursion ends. 
Not the best way to do this, but well, I cannot rewrite the code; I am afraid I will break it. maybe we need to export the `setSelectedRowIds` and use that function directly when setting the IDs in `checboxselectcolumn`. I am not sure if that is a good solution though...

![image](https://user-images.githubusercontent.com/2896170/96883634-ebd66d80-1480-11eb-8fbf-337adb184421.png)

ghiscoding commented:
I think I got it all working, I still have cleanup to do later but please note the following 2 changes (make sure you have them both before testing further).

1. I changed what the `onSelectedRowIdsChanged` event is returning, I added `filteredIds` property which is the one to use
```js
onSelectedRowIdsChanged.notify({
          "grid": grid,
          "ids": selectedRowIds,
          "filteredIds": selectedFilteredRowIds,
          "dataView": self
}, new Slick.EventData(), self);
```

2. I am now making sure inside the DataView that the `CheckboxSelectColumn` plugin has the option `applySelectOnAllPages` set to `true`  for it to trigger the `onSelectedRowIdsChanged` (if user doesn't have it enabled, then that user will keep using the grid `onSelectedRowChanged` event)
```js
var checkboxPlugin = grid.getPluginByName('CheckboxSelectColumn');
if (preserveHiddenOnSelectionChange && checkboxPlugin && checkboxPlugin.getOptions) {
  var checkboxPluginOptions = checkboxPlugin.getOptions() || {};
  if (checkboxPluginOptions.applySelectOnAllPages) {
    triggerSelectedRowIdsChange(selectedRowIds);
  }
}
```

Lastly, if I can validate that everything is working as it should, then my next step will be to cleanup some of the code. I think I can get rid of `getAllSelectedFilteredIds`, `getAllSelectedFilteredItems` and also possibly that new property `filteredIds` (I only added it for testing at this point).

arashdalir commented:
sorry mate, your change broke one part again. either I am really bad at fetching your code (which I don't believe :D I am), or the code is not working as expected.
to test, please filter your values with `value1`, select all of them, then remove the filter. at this point, you will see all the entries, and those with `value1` are selected. 
now filter with `value2` and select all of them; then remove your filter. 
now we expect to see `value1` and `value2`s to be selected, but only `value2`s are selected!

please refer to the GIF based on the changed example file. the same issue exists in my own dev-env.

![select-all-issue](https://user-images.githubusercontent.com/2896170/96896699-1ed32e00-148e-11eb-8f32-2b2b09398f69.gif)

the problem is that the `setSelectedId`, the way it's implemented at the moment, **overwrites** the previously selected values; which might make sense based on the function name, but won't make any sense for this use-case. for this case, we need to merge the newly added `selectedIds` to the list of previously selected IDs, which is something done in `grid.onSelectedRowsChanged.subscribe` function - around line 1137. we either need to let the `notify`ed event in the previous implementation of `setSelectedIds` get us to that point in code or move that portion to this location, which I'd say is a bad idea.
Principally, what you have done at this point is to circumvent the `subscribe`d event handler and to set the `selectedIds` manually, which will then cause inconsistencies if the `subscribe`d function is called from another execution path or well, for the time being is causing inconsistencies when the function `setSelectedIds` is called directly...

ghiscoding commented:
I noticed the "Select All" is also not being selected after filtering and I just fixed it, it looks like I still need `getAllSelectedFilteredIds` in Angular-Slickgrid so I'll keep all these functions

However I disagree with what you said, if you filter value "1" then Select All and then change to value "2" and Select All, I do assume (as a regular user) that Selecting All will only (and always be) the Select All of last filtered data shown on screen, so Yes I assume it will overwrite the previous selection because value "1" is no longer shown and should be discarded in new selection.

arashdalir commented:
> However I disagree with what you said,...

I strongly disagree. 

1. first of all, the way you have solved it, is technically not correct, because as said before, you have now caused all the event listeners to miss this change of `selectedIds`. 
2. ahm! no?! selecting item2 doesn't mean I don't need item1? if you select all of item1, and want to select all of item2, what should you do? write a complex filter that filters both items? what if you filter item1 and item2, and then realize you need item3 too? what if the list of items is really long and you cannot create a user-friendly multiselect dropdown (like for 1.000.000 values)? there are so many different reasons why someone might want this!
3. you are forcing users your decision! "I don't believe you wanted item1, so I selected all item2 s, although you never manually unselected item1s." you get my point? **the user** _should be able_ to decide if they want to select something in addition to previous items or instead of them. you are forcing your opinion to them, and with the current implementation of `slick-grid`, there is no nice way to allow it the other way around!
4. you cannot create negation functionality with your solution! let's assume I want all A1s, which aren't B2s. with my suggestion, I will filter all A1s and select them, then filter B2s and unselect them! now I have A1-B2! please tell me how I can achieve this with your suggestion...
5. I literally showed you our own use-case in previous screenshots and why we need to filter for various values; which as actually based on our customer's / i.e. "regular user" - requirement. so, now! your assumption that a regular user will not want this is wrong.


> noticed the "Select All" is also not being selected after filtering and I just fixed it,

I mentioned this here: https://github.com/6pac/SlickGrid/pull/545#issuecomment-714509949 
(now about the new issue:
as we know, HTML's checkbox only has 2 modes...)

have to check your code to see how your solution works. I can do it tomorrow though...

arashdalir commented:
@ghiscoding 

One last argument why you are wrong:

     * @param preserveHiddenOnSelectionChange {Boolean} Whether to keep selected items

So, no! It's not as you believe! Sorry mate!

ghiscoding commented:
I thought the default behavior was what I said but after trying some of the live Examples, I see that it's like you said, so go ahead and work on a fix if you can, I don't have more time to pass on this for the next couple days.

6pac commented:
Is this still in progress, or shall wel close it?

ghiscoding commented:
@6pac can you please leave it open, I might still come back to it someday

arashdalir commented:
I have already implemented my own solution based on my suggestion. the issue is probably still open on the main branch and this is actually an important issue, too (like the multi-sort thing); so I'd suggest we don't close it and work on a solution...

ghiscoding commented:
@arashdalir it would be better if you share your solution and try to get this merged

arashdalir commented:
> @arashdalir it would be better if you share your solution and try to get this merged

I'd really like to do so, but my changes are kind of not exactly clean and I need to go back a few revisions and try to figure it out. will get back to you on this...

arashdalir commented:
I have resolved row selection issues in [this revision](https://github.com/arashdalir/SlickGrid/tree/gemik-stable). unfortunately, it's a bit harder for me to separate the code into different smaller pieces.

basically, the following parts have to be changed:

1. `DataView` needs to change; new events and functions are added to allow the `dataview` to track the list of selected items over the filtered dataset and all data.
1. `CheckboxSelectColumn` needs to allow the selection on all pages and usage of dataview in its module, so it can interact with dataview when the data is filtered or when the user selects or unselects rows

6pac commented:
How about creating a pull request and I'll look through the changes? I think I'd probably split it up into smaller commits, but that would be a good way to at least look at the scope of the changes.
